File: app/src/main/java/com/bukonudakonusalim/takenotes/ui/main/MainActivity.java
package com.bukonudakonusalim.takenotes.ui.main;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.view.ViewCompat;
import androidx.databinding.DataBindingUtil;
import androidx.viewpager2.widget.ViewPager2;

import android.content.Intent;
import android.graphics.drawable.GradientDrawable;
import android.os.Bundle;
import android.view.View;
import android.view.inputmethod.EditorInfo;

import com.bukonudakonusalim.takenotes.data.model.NotebookModel;
import com.bukonudakonusalim.takenotes.utils.ColorUtils;
import com.bukonudakonusalim.takenotes.utils.DatabaseController;
import com.bukonudakonusalim.takenotes.R;
import com.bukonudakonusalim.takenotes.databinding.ActivityMainBinding;
import com.bukonudakonusalim.takenotes.ui.logs.LogsActivity;
import com.bukonudakonusalim.takenotes.ui.newnotebook.CreateNotebookActivity;
import com.bukonudakonusalim.takenotes.ui.notebook.NotebookActivity;
import com.bukonudakonusalim.takenotes.utils.TimeUtils;

import java.util.List;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    private ActivityMainBinding mBinding;

    private NotebooksAdapter mNotebooksAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mBinding = DataBindingUtil.setContentView(this, R.layout.activity_main);

        mBinding.btnLogs.setOnClickListener(this);
        mBinding.tvTodayDay.setText(TimeUtils.getTodayLong());
        mBinding.etNameOfPerson.setOnFocusChangeListener((view, b) -> mBinding.etNameOfPerson.setCursorVisible(b));
        mBinding.etNameOfPerson.setOnEditorActionListener((textView, i, keyEvent) -> {
            if (i == EditorInfo.IME_ACTION_DONE) {
                mBinding.etNameOfPerson.clearFocus();
                mBinding.etNameOfPerson.setCursorVisible(false);
            }
            return false;
        });
        initNotebooksViewpager();
    }

    private void initNotebooksViewpager() {
        mNotebooksAdapter = new NotebooksAdapter();
        setNotebooksViewpager();
    }

    private void setNotebooksViewpager() {
        mNotebooksAdapter.setOnNotebooksClickListener(new NotebooksAdapter.OnNotebooksClickListener() {
            @Override
            public void onNotebookClick(int pos) {
                Intent notebookScreen = new Intent(MainActivity.this, NotebookActivity.class);
                notebookScreen.putExtra("notebook_id", mNotebooksAdapter.getItemAt(pos).getId());
                startActivity(notebookScreen);
            }

            @Override
            public void onCreateClick() {
                Intent createNotebook = new Intent(MainActivity.this, CreateNotebookActivity.class);
                startActivity(createNotebook);
            }
        });

        mBinding.vpNotebooks.setAdapter(mNotebooksAdapter);
        mBinding.vpNotebooks.setOrientation(ViewPager2.ORIENTATION_HORIZONTAL);
        mBinding.vpNotebooks.setOffscreenPageLimit(3);
        setPageTransformer();
    }

    private void setPageTransformer() {
        float pageMargin = getResources().getDimensionPixelOffset(R.dimen.pageMargin);
        float selectedHeight = getResources().getDimensionPixelOffset(R.dimen.selectedHeight);
        float pageOffset = getResources().getDimensionPixelOffset(R.dimen.offset);

        mBinding.vpNotebooks.setPageTransformer((page, position) -> {
            float offset = position * -(2 * pageOffset + pageMargin);
            float height = Math.abs(position) * selectedHeight;
//            float scale = (float) (1f - (0.1 * Math.pow(position, 2)));
            if (mBinding.vpNotebooks.getOrientation() == ViewPager2.ORIENTATION_HORIZONTAL) {
                if (ViewCompat.getLayoutDirection(mBinding.vpNotebooks) == ViewCompat.LAYOUT_DIRECTION_RTL) {
                    page.setTranslationX(-offset);
                } else {
                    page.setTranslationX(offset);
                }
                page.setTranslationY(height);
//                page.setScaleY(scale);
            } else {
                page.setTranslationY(offset);
            }
        });
    }

    @Override
    protected void onResume() {
        new Thread(() -> {
            List<NotebookModel> notebooks = NotebookModel.getAllNotebooks(DatabaseController.getInstance(MainActivity.this));
            runOnUiThread(() -> {
                mNotebooksAdapter.setNotebooks(notebooks);
                mBinding.vpNotebooks.registerOnPageChangeCallback(new ViewPager2.OnPageChangeCallback() {
                    @Override
                    public void onPageSelected(int position) {
                        super.onPageSelected(position);
                        if (position != mNotebooksAdapter.getItemCount() - 1) {
                            mBinding.background.setBackground(new GradientDrawable(GradientDrawable.Orientation.TOP_BOTTOM, ColorUtils.getColorTwist(MainActivity.this, mNotebooksAdapter.getItemAt(position).getNotebookColor())));
                        } else {
                            mBinding.background.setBackground(new GradientDrawable(GradientDrawable.Orientation.TOP_BOTTOM, ColorUtils.getColorTwist(MainActivity.this, "light_blue")));
                        }
                    }
                });
            });
        }).start();
        super.onResume();
    }

    @Override
    public void onClick(View view) {
        if (view.getId() == mBinding.btnLogs.getId()) {
            Intent logs = new Intent(MainActivity.this, LogsActivity.class);
            startActivity(logs);
        }
    }
}

File: app/src/main/java/com/bukonudakonusalim/takenotes/ui/main/MainActivity.kt
package com.bukonudakonusalim.takenotes.ui.main

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.view.inputmethod.EditorInfo
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.ViewModelProvider
import androidx.viewpager2.widget.ViewPager2
import com.bukonudakonusalim.takenotes.R
import com.bukonudakonusalim.takenotes.data.DataHolder
import com.bukonudakonusalim.takenotes.data.ViewModelProviderFactory
import com.bukonudakonusalim.takenotes.data.model.Setting
import com.bukonudakonusalim.takenotes.data.viewmodel.MainViewModel
import com.bukonudakonusalim.takenotes.data.viewmodel.NotebooksViewModel
import com.bukonudakonusalim.takenotes.databinding.ActivityMainBinding
import com.bukonudakonusalim.takenotes.ui.main.NotebooksAdapter.OnNotebooksClickListener
import com.bukonudakonusalim.takenotes.ui.newnotebook.CreateNotebookActivity
import com.bukonudakonusalim.takenotes.ui.notebook.NotebookActivity
import com.bukonudakonusalim.takenotes.utils.TimeUtils
import logme.log.Logme


class MainActivity : AppCompatActivity(), View.OnClickListener {

    private lateinit var mBinding: ActivityMainBinding
    private val mMainViewModel: MainViewModel by lazy { ViewModelProvider(this, ViewModelProviderFactory(this.application, null)).get(MainViewModel::class.java) }
    private val mNotebooksViewModel: NotebooksViewModel by lazy { ViewModelProvider(this, ViewModelProviderFactory(this.application, null)).get(NotebooksViewModel::class.java) }

    private lateinit var mNotebooksAdapter: NotebooksAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Logme.wtf("Main activity started...")
        mBinding = DataBindingUtil.setContentView(this, R.layout.activity_main)

        mBinding.tvTodayDay.text = TimeUtils.getTodayLong()

        mMainViewModel.getSetting("user_name").observe(this, { setting ->
            val username: String? = setting?.value
            mBinding.etNameOfPerson.apply {
                mBinding.etNameOfPerson.setText(username ?: "!")
                setOnFocusChangeListener { _, b -> mBinding.etNameOfPerson.isCursorVisible = b }
                setOnEditorActionListener { _, i, _ ->
                    if (i === EditorInfo.IME_ACTION_DONE) {
                        mMainViewModel.insertSetting(Setting("user_name", mBinding.etNameOfPerson.text.toString().trim()))
                        mBinding.etNameOfPerson.apply {
                            setText(mBinding.etNameOfPerson.text.toString().trim())
                            clearFocus()
                            isCursorVisible = false
                        }
                    }
                    false
                }
            }
        })

        initNotebooksViewpager()
    }

    private fun initNotebooksViewpager() {
        mNotebooksAdapter = NotebooksAdapter()
        setNotebooksViewpager()
    }

    private fun setNotebooksViewpager() {
        mNotebooksViewModel.allNotebooks.observe(this, { notebooks ->
            mNotebooksAdapter.setNotebooks(notebooks)
        })
        mNotebooksAdapter.setOnNotebooksClickListener(object : OnNotebooksClickListener {
            override fun onNotebookClick(pos: Int) {
                val notebookScreen = Intent(this@MainActivity, NotebookActivity::class.java)
                notebookScreen.putExtra("notebook_id", mNotebooksAdapter.getItemAt(pos).id)
                startActivity(notebookScreen)
            }

            override fun onCreateClick() {
                val createNotebook = Intent(this@MainActivity, CreateNotebookActivity::class.java)
                startActivity(createNotebook)
            }
        })
        mBinding.vpNotebooks.adapter = mNotebooksAdapter
        mBinding.vpNotebooks.orientation = ViewPager2.ORIENTATION_HORIZONTAL
        mBinding.vpNotebooks.offscreenPageLimit = 3
        setPageTransformer()
    }

    private fun setPageTransformer() {
        val pageMargin = resources.getDimensionPixelOffset(R.dimen.pageMargin).toFloat()
        val selectedHeight = resources.getDimensionPixelOffset(R.dimen.selectedHeight).toFloat()
        val pageOffset = resources.getDimensionPixelOffset(R.dimen.offset).toFloat()
        mBinding.vpNotebooks.setPageTransformer { page, position ->
            val offset = position * -(2 * pageOffset + pageMargin)
            val height = Math.abs(position) * selectedHeight
            //            float scale = (float) (1f - (0.1 * Math.pow(position, 2)));
            if (mBinding.vpNotebooks.orientation === ViewPager2.ORIENTATION_HORIZONTAL) {
                if (ViewCompat.getLayoutDirection(mBinding.vpNotebooks) == ViewCompat.LAYOUT_DIRECTION_RTL) {
                    page.translationX = -offset
                } else {
                    page.translationX = offset
                }
                page.translationY = height
                //                page.setScaleY(scale);
            } else {
                page.translationY = offset
            }
        }
    }

    override fun onClick(view: View?) {
        when (view?.id) {
            //
        }
    }
}